src/plugins/expressions/common/ast/ast_builder.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/* eslint-disable max-classes-per-file */
+
+import { ExpressionAstArgument, ExpressionAstExpression, ExpressionAstFunction } from './types';
+
+const createExpression = (chain: ExpressionAstFunction[]): ExpressionAstExpression => {
+  return {
+    type: 'expression',
+    chain,
+  };
+};
+
+const createFunction = (
+  name: string,
+  args: Record<string, ExpressionAstArgument[] | ExpressionAstArgument> = {}
+): ExpressionAstFunction => {
```

Comment:
> how can you enforce that all provided args are present on the function when they aren't added all at once?

I guess I should've rephrased this to "you can't enforce that all provided args are present on the function". So if we let folks update functions with an individual argument instead of passing all of them at once, the tradeoff we are making is that we aren't 100% sure all args have been provided.

The only way around this I can think of would be to force users to provide a full list of the required arguments when instantiating the function builder, and then allow them to manipulate the args one-by-one later. That way you'd know that no required args are missing.

Which brings another question to mind - is there a use case for _removing_ an arg from a function?

Reply:
the ability to add arguments later and remove them just makes the api way easier to consume (and refacture current string concatenation logic). however i think everything should still be possible with the above mentioned limitations (you need to provide default values for all required arguments and you cannot remove arguments).

what we are discussing here is are we willing to trade typesafety for better usability ? (we could still provide runtime checks if we opt for not being type safe here.

Reply:
> you need to provide default values for all required arguments and you cannot remove arguments

Agreed, IMHO this is an acceptable tradeoff for type safety vs useability. If you are forced to provide defaults for all required args up front, then you have the safety of knowing that you aren't missing anything. And then any subsequent call to update one of those args will enforce the correct type.

Reply:
We could also try to do runtime checks, the only tricky part there is that it would probably require looking up the referenced functions from the registry to get their metadata, vs the current implementation which is much simpler in that it is static.